### ISSUE TITLE
Reproducing performance on single GPU

### ISSUE THREAD

**Issue by szatidani** [open-mmlab/mmsegmentation#179]
I've started to work with this awesome repo for semantic segmentation. I'd like to reproduce some of the results of the available models as a first step. I'm working with DeepLabV3+ and PSPNet right now. 

I've done a training using PSPNet on Cityscapes dataset with this config: `pspnet_r50-d8_512x1024_80k_cityscapes.py` having the SyncBN changed to BN for single GPU training. According to the benchmarks on [PSPNet](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/pspnet), it supposed to reach **mIoU=79.76**, but I only got **mIoU=69.13** after 80k of iterations.

I understand that the provided checkpoints have been trained on a 4 GPU system, so the config parameters are also based on that setup.  In the [documentation](https://mmsegmentation.readthedocs.io/en/latest/getting_started.html#train-a-model), I've noticed the note: 

> **Important**: The default learning rate in config files is for 4 GPUs and 2 img/gpu (batch size = 4x2 = 8). Equivalently, you may also use 8 GPUs and 1 imgs/gpu since all models using cross-GPU SyncBN.

**My question:** Is there **a way to reproduce** the reported results **on a single GPU** system? (I have a single Tesla P100 with 16 GB) If so, **what parameters are suggested?** 

Based on the documentation, I assume **learning rate, batch size and iterations should be changed**? Do you have **suggested values**  of those to be able to **reproduce** your results? Also, ****Am I missing** any **important parameter** that could help?**

I really appreciate the help.
@xvjiarui

**Comment by xvjiarui** [comment]
Hi @szatidani 
It will take longer time. Since you have 16Gb memory, you may try batch_size*2, learning_rate*0.5, iterations*2.

**Comment by caodroid** [comment]
Hi @xvjiarui 
I also meet the same problem with this friend, and  ''batch_size2, learning_rate0.5, iterations*2.'' means double batch_size half learning_rate or setting batch_size= 2 learning_rate=0.5？ thank you for your reply！

**Comment by atiszabo** [comment]
@xvjiarui thank you for the answers, although as @caodroid pointed out there is a bit of an ambiguity that would be nice to clarify. 

- Also, a bit about the batch size: the default config is 2 / gpu, doubling it, would still only mean 4 whereas the authors of DeepLabV3+ recommend higher, possible at least 8 to be able to train batch norm training. In the 4 GPU configuration, it is met with SyncBN, which makes me think that is there way to simulate 4 GPU system on a single one and do similar trick as SynBN? So as a naive idea, do 4 times a propagation with the batch of 2 on one machine and then update similarly as it is done on 4 gpus. Then I think with longer iteration set and possibly adjusted learning rate maybe there is a higher chance to reproduce the results. **Would that be feasable?**

- For one training I could initiate a 4x P100 GPU system and train with **default settings** from 'deeplabv3plus_r101-d8_769x769_80k_cityscapes.py', here is what I got:
`Scope                 mIoU       mAcc       aAcc`
`global               80.04      88.46      96.53`
which is now much closer to the reported mIoU of **80.04** vs. **80.98**, but still a noticeable difference. **So my questions are:** Is the reported number an average from multiple runs? In that case do you have standard deviation calculated? Is such stochastic behavior expected from the framework?

**Comment by hjraad** [comment]
@xvjiarui @hellock : the questions above from @atiszabo are relevant to several users, seems like reproduction on single GPU is a challenge. Could you please elaborate?

**Comment by sumenpuyuan** [comment]
反过来讲，多卡， syncbn这么有效的吗……

**Comment by hjraad** [comment]
> Hi @szatidani
> It will take longer time. Since you have 16Gb memory, you may try batch_size_2, learning_rate_0.5, iterations*2.

Does it only take longer, or reach an inferior performance?

**Comment by caodroid** [comment]
@hjraad Hi，I retrained the DANet :dan_r50_d8_512x1024_80k_cityscaples.py on my mechine with 2 Tesla P40 GPU, and batch_size_4per gpu, learning_rate=0.5,  the result is """mIoU: 0.7981, mAcc: 0.8662, aAcc: 0.9625""" , a little better than the  reference. as for training time, I spent about 31h.  I alse see the trick  to adjust batch_size learning_rate in mmdetection documentation  follows  ：
*Important*: The default learning rate in config files is for 8 GPUs and 2 img/gpu (batch size = 8*2 = 16). According to the Linear Scaling Rule, you need to set the learning rate proportional to the batch size if you use different GPUs or images per GPU, e.g., lr=0.01 for 4 GPUs * 2 img/gpu and lr=0.08 for 16 GPUs * 4 img/gpu.  [https://mmdetection.readthedocs.io/en/latest/getting_started.html#train-a-model](url)
I used it in my retraining, it's really works. hope will help you.

**Comment by xvjiarui** [comment]
> @xvjiarui thank you for the answers, although as @caodroid pointed out there is a bit of an ambiguity that would be nice to clarify.
> 
> * Also, a bit about the batch size: the default config is 2 / gpu, doubling it, would still only mean 4 whereas the authors of DeepLabV3+ recommend higher, possible at least 8 to be able to train batch norm training. In the 4 GPU configuration, it is met with SyncBN, which makes me think that is there way to simulate 4 GPU system on a single one and do similar trick as SynBN? So as a naive idea, do 4 times a propagation with the batch of 2 on one machine and then update similarly as it is done on 4 gpus. Then I think with longer iteration set and possibly adjusted learning rate maybe there is a higher chance to reproduce the results. **Would that be feasable?**
> * For one training I could initiate a 4x P100 GPU system and train with **default settings** from 'deeplabv3plus_r101-d8_769x769_80k_cityscapes.py', here is what I got:
>   `Scope                 mIoU       mAcc       aAcc`
>   `global               80.04      88.46      96.53`
>   which is now much closer to the reported mIoU of **80.04** vs. **80.98**, but still a noticeable difference. **So my questions are:** Is the reported number an average from multiple runs? In that case do you have standard deviation calculated? Is such stochastic behavior expected from the framework?

Sorry for the late reply. It is not an average of multiple runs.It is normal to oscillate ~0.5 for Deeplab based model, ~0.7 for attention based model.

**Comment by hjraad** [comment]
@xvjiarui : thanks for your reaction. I believe this shouldn't be closed before we reach a consensus:

==>  Good points, (due to the batch norm?) an oscillation is the final result up to 0.5 mIoU (which is large btw), is acceptable. I experienced the same with DeepLabv3+ too. 

Now points still unclear:
==> The results you report at the best outta the bunch right? No averaging according to your response to @atiszabo. 
==> Did you apply the recommendation by @caodroid on the documentation? He might have raised a good point there.
==> Do you agree with the explanations @caodroid provided on the relationship between learning, batch size and number of GPUs? You mentioned run time (iterations) will play a role. Could you please share a rule of thumb hint on this, and also maybe another good addition to your documentation. I'm sure people stumble upon this, so going to be a very useful hint. 
Many thanks; Looking Forward!

**Comment by atiszabo** [comment]
I pushed a run to test this hypothesis about being able to reproduce the reported results with a single gpu. I've adjusted the parameters based on the mentioned Linear scaling rule by @caodroid. 

The parameters, instead of 4 gpus -> 1: so new_lr = orig_lr / 4. Unfortunately, the **batch size can't be doubled on a single machine** @xvjiarui, but still I made the iterations longer based on your suggestions.

**Params**:
lr = 0.0025, (instead of 0.01)
max_iter =  320k, (instead of 80k)

Around 160k, it reached miou of 70 on cityscapes, but it keeps oscillating since 40k, around miou of 64(+-5). With a batch_size of 2 the gradients seems very noisy to update the model properly.
Any suggestions?

**Comment by atiszabo** [comment]
Another update on the single gpu performance. We lowered the lr even further to try to mitigate the noisy batch, also extended the number of iterations.
**Params**:

- lr= 0.0005
- for lr_config: min_lr=5e-6
- total_iters = 640k

Until it reaches **miou~=70** it goes up steadily and does not oscillate a lot. After that it starts to go back and forth in performance **between miou~=65 and miou~=73**, but it can't pass that max value at any point for long number of iterations.

It seems it is impossible to reach the same miou~80 regime with single gpu with current implementation.
